File: fit-zig/src/runner.zig
const std = @import("std");
const repo = @import("repo.zig");
const main = @import("main.zig");

const MAX_REPO_NAME_WIDTH: usize = 24;

/// Format repo name with fixed width: truncate long names, pad short ones
pub fn formatRepoName(allocator: std.mem.Allocator, name: []const u8) ![]const u8 {
    var display_name: []const u8 = undefined;
    var allocated = false;

    if (name.len > MAX_REPO_NAME_WIDTH) {
        // Truncate: "long-name-..." (20 chars + "-...")
        const truncated = try std.fmt.allocPrint(allocator, "{s}-...", .{name[0 .. MAX_REPO_NAME_WIDTH - 4]});
        display_name = truncated;
        allocated = true;
    } else {
        display_name = name;
    }
    defer if (allocated) allocator.free(display_name);

    // Format with padding: "[name                    ]"
    return std.fmt.allocPrint(allocator, "[{s: <24}]", .{display_name});
}

/// Execution context holding configuration for running git commands
pub const ExecutionContext = struct {
    workers: usize,
    dry_run: bool,
    url_scheme: ?main.UrlScheme,

    pub fn init(workers: usize, dry_run: bool, url_scheme: ?main.UrlScheme) ExecutionContext {
        return .{
            .workers = workers,
            .dry_run = dry_run,
            .url_scheme = url_scheme,
        };
    }

    /// Get git config args for URL scheme rewriting
    pub fn urlSchemeArgs(self: *const ExecutionContext) []const []const u8 {
        return switch (self.url_scheme orelse return &.{}) {
            .ssh => &.{ "-c", "url.git@github.com:.insteadOf=https://github.com/" },
            .https => &.{ "-c", "url.https://github.com/.insteadOf=git@github.com:" },
        };
    }
};

/// A git command ready to be executed against a repository
pub const GitCommand = struct {
    repo_path: []const u8,
    args: []const []const u8,
    url_scheme_args: []const []const u8,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator, repo_path: []const u8, args: []const []const u8, url_scheme_args: []const []const u8) GitCommand {
        return .{
            .repo_path = repo_path,
            .args = args,
            .url_scheme_args = url_scheme_args,
            .allocator = allocator,
        };
    }

    /// Build the command string for display (used in dry-run and errors)
    pub fn commandString(self: *const GitCommand) ![]const u8 {
        var parts: std.ArrayList(u8) = .empty;
        errdefer parts.deinit(self.allocator);

        try parts.appendSlice(self.allocator, "git");

        // Add URL scheme args (e.g., -c url.git@github.com:...)
        for (self.url_scheme_args) |arg| {
            try parts.append(self.allocator, ' ');
            try parts.appendSlice(self.allocator, arg);
        }

        try parts.appendSlice(self.allocator, " -C ");
        try parts.appendSlice(self.allocator, self.repo_path);

        for (self.args) |arg| {
            try parts.append(self.allocator, ' ');
            try parts.appendSlice(self.allocator, arg);
        }

        return parts.toOwnedSlice(self.allocator);
    }

    /// Execute the git command
    pub fn execute(self: *const GitCommand) CommandResult {
        // Build argv for child process
        var argv: std.ArrayList([]const u8) = .empty;
        defer argv.deinit(self.allocator);

        argv.append(self.allocator, "git") catch return .{ .err = .{ .repo_name = repo.repoName(self.repo_path), .message = "failed to build argv" } };
        // Add URL scheme args (e.g., -c url.git@github.com:...)
        for (self.url_scheme_args) |arg| {
            argv.append(self.allocator, arg) catch return .{ .err = .{ .repo_name = repo.repoName(self.repo_path), .message = "failed to build argv" } };
        }
        argv.append(self.allocator, "-C") catch return .{ .err = .{ .repo_name = repo.repoName(self.repo_path), .message = "failed to build argv" } };
        argv.append(self.allocator, self.repo_path) catch return .{ .err = .{ .repo_name = repo.repoName(self.repo_path), .message = "failed to build argv" } };
        for (self.args) |arg| {
            argv.append(self.allocator, arg) catch return .{ .err = .{ .repo_name = repo.repoName(self.repo_path), .message = "failed to build argv" } };
        }

        var child = std.process.Child.init(argv.items, self.allocator);
        child.stdout_behavior = .Pipe;
        child.stderr_behavior = .Pipe;

        child.spawn() catch |e| {
            return .{ .err = .{
                .repo_name = repo.repoName(self.repo_path),
                .message = @errorName(e),
            } };
        };

        // Read all output using the File's readToEndAlloc
        const stdout = child.stdout.?.readToEndAlloc(self.allocator, 1024 * 1024) catch |e| {
            return .{ .err = .{
                .repo_name = repo.repoName(self.repo_path),
                .message = @errorName(e),
            } };
        };

        const stderr = child.stderr.?.readToEndAlloc(self.allocator, 1024 * 1024) catch |e| {
            self.allocator.free(stdout);
            return .{ .err = .{
                .repo_name = repo.repoName(self.repo_path),
                .message = @errorName(e),
            } };
        };

        const term = child.wait() catch |e| {
            self.allocator.free(stdout);
            self.allocator.free(stderr);
            return .{ .err = .{
                .repo_name = repo.repoName(self.repo_path),
                .message = @errorName(e),
            } };
        };

        const success = term.Exited == 0;

        return .{ .executed = .{
            .repo_name = repo.repoName(self.repo_path),
            .stdout = stdout,
            .stderr = stderr,
            .success = success,
        } };
    }
};

/// Result of executing a git command
pub const CommandResult = union(enum) {
    dry_run: []const u8,
    executed: struct {
        repo_name: []const u8,
        stdout: []const u8,
        stderr: []const u8,
        success: bool,
    },
    err: struct {
        repo_name: []const u8,
        message: []const u8,
    },
};

/// Output formatter function type
/// Formatters accept an allocator and return owned memory that the caller must free
pub const OutputFormatter = *const fn (allocator: std.mem.Allocator, stdout: []const u8, stderr: []const u8, success: bool) error{OutOfMemory}![]const u8;

/// Thread context for parallel execution
const ThreadContext = struct {
    repos: []const []const u8,
    start_idx: usize,
    end_idx: usize,
    extra_args: []const []const u8,
    formatter: OutputFormatter,
    ctx: *const ExecutionContext,
    stdout_mutex: *std.Thread.Mutex,
    allocator: std.mem.Allocator,
};

fn workerThread(thread_ctx: ThreadContext) void {
    const allocator = thread_ctx.allocator;
    const url_scheme_args = thread_ctx.ctx.urlSchemeArgs();

    for (thread_ctx.start_idx..thread_ctx.end_idx) |i| {
        const repo_path = thread_ctx.repos[i];
        const cmd = GitCommand.init(allocator, repo_path, thread_ctx.extra_args, url_scheme_args);

        var output_line: []const u8 = undefined;
        var needs_free = true;

        if (thread_ctx.ctx.dry_run) {
            if (cmd.commandString()) |cmd_str| {
                output_line = cmd_str;
            } else |_| {
                output_line = "error building command";
                needs_free = false;
            }
        } else {
            const result = cmd.execute();

            switch (result) {
                .dry_run => |cmd_str| {
                    output_line = cmd_str;
                },
                .executed => |exec| {
                    // Get formatted output (or fallback)
                    var formatted: []const u8 = undefined;
                    var formatted_allocated = false;
                    if (thread_ctx.formatter(allocator, exec.stdout, exec.stderr, exec.success)) |f| {
                        formatted = f;
                        formatted_allocated = true;
                    } else |_| {
                        formatted = "format error";
                    }

                    const repo_name = repo.repoName(repo_path);
                    var name_formatted: []const u8 = undefined;
                    var name_allocated = false;
                    if (formatRepoName(allocator, repo_name)) |n| {
                        name_formatted = n;
                        name_allocated = true;
                    } else |_| {
                        name_formatted = "[???]";
                    }

                    // Build final output line
                    if (std.fmt.allocPrint(allocator, "{s} {s}", .{ name_formatted, formatted })) |line| {
                        output_line = line;
                    } else |_| {
                        output_line = "format error";
                        needs_free = false;
                    }

                    // Free intermediate allocations
                    if (formatted_allocated) allocator.free(formatted);
                    if (name_allocated) allocator.free(name_formatted);

                    // Free stdout/stderr from command execution
                    allocator.free(exec.stdout);
                    allocator.free(exec.stderr);
                },
                .err => |e| {
                    const repo_name = repo.repoName(repo_path);
                    var name_formatted: []const u8 = undefined;
                    var name_allocated = false;
                    if (formatRepoName(allocator, repo_name)) |n| {
                        name_formatted = n;
                        name_allocated = true;
                    } else |_| {
                        name_formatted = "[???]";
                    }

                    if (std.fmt.allocPrint(allocator, "{s} ERROR: {s}", .{ name_formatted, e.message })) |line| {
                        output_line = line;
                    } else |_| {
                        output_line = "format error";
                        needs_free = false;
                    }

                    if (name_allocated) allocator.free(name_formatted);
                },
            }
        }

        // Lock stdout and print atomically
        thread_ctx.stdout_mutex.lock();
        const stdout_file = std.fs.File.stdout();
        stdout_file.writeAll(output_line) catch {};
        stdout_file.writeAll("\n") catch {};
        thread_ctx.stdout_mutex.unlock();

        // Free this iteration's allocation
        if (needs_free) {
            allocator.free(output_line);
        }
    }
}

/// Run commands in parallel across all repos
pub fn runParallel(
    allocator: std.mem.Allocator,
    ctx: *const ExecutionContext,
    repos: []const []const u8,
    extra_args: []const []const u8,
    formatter: OutputFormatter,
) !void {
    if (repos.len == 0) return;

    var stdout_mutex = std.Thread.Mutex{};

    // Calculate work distribution
    const num_workers = @min(ctx.workers, repos.len);
    const repos_per_worker = repos.len / num_workers;
    const extra_repos = repos.len % num_workers;

    var threads: std.ArrayList(std.Thread) = .empty;
    defer threads.deinit(allocator);

    var start_idx: usize = 0;
    for (0..num_workers) |worker_id| {
        // Distribute extra repos among first workers
        const worker_repos = repos_per_worker + @as(usize, if (worker_id < extra_repos) 1 else 0);
        const end_idx = start_idx + worker_repos;

        const thread_ctx = ThreadContext{
            .repos = repos,
            .start_idx = start_idx,
            .end_idx = end_idx,
            .extra_args = extra_args,
            .formatter = formatter,
            .ctx = ctx,
            .stdout_mutex = &stdout_mutex,
            .allocator = allocator,
        };

        const thread = try std.Thread.spawn(.{}, workerThread, .{thread_ctx});
        try threads.append(allocator, thread);

        start_idx = end_idx;
    }

    // Wait for all threads to complete
    for (threads.items) |thread| {
        thread.join();
    }
}

test "formatRepoName short name" {
    const allocator = std.testing.allocator;
    const result = try formatRepoName(allocator, "my-repo");
    defer allocator.free(result);
    // 26 chars total: [ + 24 chars padded + ]
    try std.testing.expectEqualStrings("[my-repo                 ]", result);
    try std.testing.expect(result.len == 26);
}

test "formatRepoName truncated" {
    const allocator = std.testing.allocator;
    const result = try formatRepoName(allocator, "this-is-a-very-long-repository-name");
    defer allocator.free(result);
    // Truncated to 20 chars + "-..." = 24 chars, wrapped in brackets
    try std.testing.expectEqualStrings("[this-is-a-very-long--...]", result);
    try std.testing.expect(result.len == 26);
}

File: fit-rust/src/runner.rs
use anyhow::Result;
use std::path::PathBuf;
use std::process::{Command, Output, Stdio};
use std::sync::mpsc;
use std::sync::{Arc, Condvar, Mutex};

use crate::repo::repo_name;

/// Simple counting semaphore using stdlib primitives.
/// Allows limiting concurrent operations to N at a time.
struct Semaphore {
    count: Mutex<usize>,
    cond: Condvar,
}

impl Semaphore {
    fn new(permits: usize) -> Self {
        Semaphore {
            count: Mutex::new(permits),
            cond: Condvar::new(),
        }
    }

    /// Acquire a permit, blocking if none available.
    fn acquire(&self) {
        let mut count = self.count.lock().unwrap();
        while *count == 0 {
            count = self.cond.wait(count).unwrap();
        }
        *count -= 1;
    }

    /// Release a permit, waking one waiting thread.
    fn release(&self) {
        let mut count = self.count.lock().unwrap();
        *count += 1;
        self.cond.notify_one();
    }
}

const MAX_REPO_NAME_WIDTH: usize = 24;

/// URL scheme to force for git operations
#[derive(Clone, Copy)]
pub enum UrlScheme {
    /// Force SSH: git@github.com:user/repo
    Ssh,
    /// Force HTTPS: https://github.com/user/repo
    Https,
}

/// Format repo name with fixed width: truncate long names, pad short ones
fn format_repo_name(name: &str) -> String {
    let display_name = if name.len() > MAX_REPO_NAME_WIDTH {
        format!("{}-...", &name[..MAX_REPO_NAME_WIDTH - 4])
    } else {
        name.to_string()
    };
    format!("[{:<width$}]", display_name, width = MAX_REPO_NAME_WIDTH)
}

/// Execution context holding configuration for running git commands
pub struct ExecutionContext {
    dry_run: bool,
    url_scheme: Option<UrlScheme>,
    max_connections: usize,
}

impl ExecutionContext {
    pub fn new(dry_run: bool, url_scheme: Option<UrlScheme>, max_connections: usize) -> Self {
        Self {
            dry_run,
            url_scheme,
            max_connections,
        }
    }

    pub fn is_dry_run(&self) -> bool {
        self.dry_run
    }

    pub fn url_scheme(&self) -> Option<UrlScheme> {
        self.url_scheme
    }

    pub fn max_connections(&self) -> usize {
        self.max_connections
    }
}

/// A git command ready to be executed against a repository
pub struct GitCommand {
    pub repo_path: PathBuf,
    pub args: Vec<String>,
}

impl GitCommand {
    pub fn new(repo_path: PathBuf, args: Vec<String>) -> Self {
        Self { repo_path, args }
    }

    /// Spawn the git command without waiting for completion.
    /// Returns immediately with a Child process handle.
    pub fn spawn(&self, url_scheme: Option<UrlScheme>) -> std::io::Result<std::process::Child> {
        let mut cmd = Command::new("git");

        // Inject URL scheme override if specified (must come before other args)
        if let Some(scheme) = url_scheme {
            match scheme {
                UrlScheme::Ssh => {
                    cmd.arg("-c")
                        .arg("url.git@github.com:.insteadOf=https://github.com/");
                }
                UrlScheme::Https => {
                    cmd.arg("-c")
                        .arg("url.https://github.com/.insteadOf=git@github.com:");
                }
            }
        }

        cmd.arg("-C")
            .arg(&self.repo_path)
            .args(&self.args)
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .env("GIT_TERMINAL_PROMPT", "0")
            .spawn()
    }

    /// Build the full command string for display (used in dry-run)
    pub fn command_string_with_scheme(&self, url_scheme: Option<UrlScheme>) -> String {
        let scheme_args = match url_scheme {
            Some(UrlScheme::Ssh) => "-c \"url.git@github.com:.insteadOf=https://github.com/\" ",
            Some(UrlScheme::Https) => "-c \"url.https://github.com/.insteadOf=git@github.com:\" ",
            None => "",
        };
        format!(
            "git {}-C {} {}",
            scheme_args,
            self.repo_path.display(),
            self.args.join(" ")
        )
    }
}

/// Trait for formatting command output into one line
pub trait OutputFormatter: Sync {
    fn format(&self, output: &Output) -> String;
}

/// Run commands in parallel across all repos with streaming output.
///
/// Results are printed in alphabetical order (repos are pre-sorted) as soon as
/// contiguous results are available. Uses head-of-line blocking: if repo "aaa"
/// is slow, "bbb" and "ccc" won't print until "aaa" completes.
///
/// Uses thread-per-process pattern with `wait_with_output()` which is deadlock-safe
/// (stdlib internally spawns threads to drain stdout/stderr concurrently).
pub fn run_parallel<F>(
    ctx: &ExecutionContext,
    repos: &[PathBuf],
    build_command: F,
    formatter: &dyn OutputFormatter,
) -> Result<()>
where
    F: Fn(&PathBuf) -> GitCommand + Sync,
{
    let url_scheme = ctx.url_scheme();

    // Handle dry-run mode
    if ctx.is_dry_run() {
        for repo in repos {
            let cmd = build_command(repo);
            println!("{}", cmd.command_string_with_scheme(url_scheme));
        }
        return Ok(());
    }

    let max_workers = ctx.max_connections();

    // Create optional semaphore for concurrency limiting
    // None when unlimited (0) or when workers >= repos
    let semaphore = if max_workers > 0 && max_workers < repos.len() {
        Some(Arc::new(Semaphore::new(max_workers)))
    } else {
        None
    };

    // Results storage: None means "not yet received"
    let mut results: Vec<Option<(PathBuf, Result<Output, std::io::Error>)>> =
        (0..repos.len()).map(|_| None).collect();
    let mut next_to_print: usize = 0;

    let (tx, rx) = mpsc::channel();

    std::thread::scope(|s| {
        // Spawn threads that send results to channel as they complete
        for (idx, repo) in repos.iter().enumerate() {
            let tx = tx.clone();
            let cmd = build_command(repo);
            let repo = repo.clone();
            let sem = semaphore.clone();

            s.spawn(move || {
                if let Some(ref sem) = sem {
                    sem.acquire();
                }

                let result = cmd.spawn(url_scheme).and_then(|c| c.wait_with_output());

                if let Some(ref sem) = sem {
                    sem.release();
                }

                let _ = tx.send((idx, repo, result));
            });
        }
        drop(tx); // Close sender so rx iterator ends when all threads complete

        // Receive and print in-order as contiguous results arrive
        for (idx, repo, result) in rx {
            results[idx] = Some((repo, result));

            // Print all contiguous completed results from the head
            while next_to_print < results.len() {
                if let Some((ref repo_path, ref res)) = results[next_to_print] {
                    print_result(repo_path, res, formatter);
                    next_to_print += 1;
                } else {
                    break;
                }
            }
        }
    });

    Ok(())
}

/// Print result for a single repository
fn print_result(
    repo_path: &std::path::Path,
    result: &Result<Output, std::io::Error>,
    formatter: &dyn OutputFormatter,
) {
    let name = repo_name(repo_path);
    let output_line = match result {
        Ok(output) => {
            let formatted = formatter.format(output);
            format!("{} {}", format_repo_name(&name), formatted)
        }
        Err(e) => format!("{} ERROR: {}", format_repo_name(&name), e),
    };
    println!("{}", output_line);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_repo_name_short() {
        let result = format_repo_name("my-repo");
        assert_eq!(result, "[my-repo                 ]");
        assert_eq!(result.len(), 26); // [ + 24 + ]
    }

    #[test]
    fn test_format_repo_name_exact_length() {
        let result = format_repo_name("exactly-twenty-four-chr");
        assert_eq!(result.len(), 26);
    }

    #[test]
    fn test_format_repo_name_truncated() {
        let result = format_repo_name("this-is-a-very-long-repository-name");
        assert_eq!(result, "[this-is-a-very-long--...]");
        assert_eq!(result.len(), 26);
    }

    /// Test that large output (>64KB) doesn't cause pipe buffer deadlock.
    /// wait_with_output() internally spawns threads to drain pipes, so this should complete.
    #[test]
    fn test_large_output_no_deadlock() {
        use std::process::Stdio;
        use std::time::{Duration, Instant};

        let start = Instant::now();
        let timeout = Duration::from_secs(5);

        // Spawn a process that outputs 100KB (more than 64KB pipe buffer)
        let child = Command::new("head")
            .args(["-c", "100000", "/dev/zero"])
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .expect("Failed to spawn head command");

        // wait_with_output handles pipe draining internally - no deadlock
        let output = child.wait_with_output().expect("Failed to wait for output");

        // Verify we got all the output
        assert_eq!(
            output.stdout.len(),
            100000,
            "Expected 100000 bytes, got {}",
            output.stdout.len()
        );

        // Verify it didn't take suspiciously long (would indicate near-deadlock)
        assert!(
            start.elapsed() < timeout,
            "Test took too long - possible deadlock: {:?}",
            start.elapsed()
        );
    }
}
